File: 056.Merge Intervals.py
# Definition for an interval.
# class Interval(object):
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class Solution(object):
    def merge(self, intervals):
        """
        :type intervals: List[Interval]
        :rtype: List[Interval]
        """
        intervals.sort(key=lambda x: x.start)
        i = 0
        while i < len(intervals) - 1:
            if self.do_merge(intervals[i], intervals[i + 1]):
                intervals.pop(i + 1)
            else:
                i += 1
        return intervals

    def do_merge(self, inv1, inv2):
        if inv2.start >= inv1.start and inv2.end <= inv1.end:
            return True
        elif inv2.start >= inv1.start and inv2.start <= inv1.end:
            inv1.end = inv2.end
            return True
        elif inv2.end >= inv1.start and inv2.end <= inv1.end:
            inv1.start = inv2.start
            return True
        return False

File: 056.Merge Intervals.java
/*
    Given a collection of intervals, merge all overlapping intervals.

    For example,
    Given [1,3],[2,6],[8,10],[15,18],
    return [1,6],[8,10],[15,18].
     */

//     Definition for an interval.
     public class Interval {
         int start;
         int end;

         Interval() {
             start = 0;
             end = 0;
         }

         Interval(int s, int e) {
             start = s;
             end = e;
         }
     }

    public List<Interval> merge(List<Interval> intervals) {
        if (intervals.size() <= 1)
            return intervals;
        for (int i=0; i<intervals.size(); i++) {
            for (int j=i+1; j<intervals.size(); j++)
                if ((intervals.get(i).end >= intervals.get(j).start && intervals.get(i).start <= intervals.get(j).start)
                    || (intervals.get(j).end >= intervals.get(i).start && intervals.get(j).start <= intervals.get(i).start)) {
                    intervals.get(i).start = Math.min(intervals.get(i).start, intervals.get(j).start);
                    intervals.get(i).end = Math.max(intervals.get(i).end, intervals.get(j).end);
                    intervals.remove(j);
                    i--; break;
                }
        }

        return intervals;
    }
